File: modyn/storage/internal/file_wrapper/single_sample_file_wrapper.py
"""A file wrapper for files that contains only one sample and metadata."""

import logging
import pathlib
from typing import Optional

from modyn.storage.internal.file_wrapper.abstract_file_wrapper import AbstractFileWrapper
from modyn.storage.internal.file_wrapper.file_wrapper_type import FileWrapperType
from modyn.storage.internal.filesystem_wrapper.abstract_filesystem_wrapper import AbstractFileSystemWrapper

logger = logging.getLogger(__name__)


class SingleSampleFileWrapper(AbstractFileWrapper):
    """A file wrapper for files that contains only one sample and metadata.

    For example, a file that contains only one image and metadata.
    The metadata is stored in a json file with the same name as the image file.
    """

    def __init__(self, file_path: str, file_wrapper_config: dict, filesystem_wrapper: AbstractFileSystemWrapper):
        """Init file wrapper.

        Args:
            file_path (str): File path
            file_wrapper_config (dict): File wrapper config
        """
        super().__init__(file_path, file_wrapper_config, filesystem_wrapper)
        self.file_wrapper_type = FileWrapperType.SingleSampleFileWrapper

    def get_number_of_samples(self) -> int:
        """Get the size of the file in number of samples.

        If the file has the correct file extension, it contains only one sample.

        Returns:
            int: Number of samples
        """
        if not self.file_path.endswith(self.file_wrapper_config["file_extension"]):
            return 0
        return 1

    def get_samples(self, start: int, end: int) -> bytes:
        """Get the samples from the file.

        Args:
            start (int): start index
            end (int): end index

        Raises:
            IndexError: If the start and end index are not 0 and 1

        Returns:
            bytes: Samples
        """
        if start != 0 or end != 1:
            raise IndexError("SingleSampleFileWrapper contains only one sample.")
        return self.get_sample(0)

    def get_sample(self, index: int) -> bytes:
        r"""Return the sample as bytes.

        Args:
            index (int): Index

        Raises:
            ValueError: If the file has the wrong file extension
            IndexError: If the index is not 0

        Returns:
            bytes: Sample
        """
        if self.get_number_of_samples() == 0:
            raise ValueError("File has wrong file extension.")
        if index != 0:
            raise IndexError("SingleSampleFileWrapper contains only one sample.")
        data_file = self.filesystem_wrapper.get(self.file_path)
        return data_file

    def get_label(self, index: int) -> Optional[int]:
        """Get the label of the sample at the given index.

        Args:
            index (int): Index

        Raises:
            ValueError: If the file has the wrong file extension
            IndexError: If the index is not 0

        Returns:
            int: Label if exists, else None
        """
        if self.get_number_of_samples() == 0:
            raise ValueError("File has wrong file extension.")
        if index != 0:
            raise IndexError("SingleSampleFileWrapper contains only one sample.")
        if (
            "label_file_extension" not in self.file_wrapper_config
            or self.file_wrapper_config["label_file_extension"] is None
        ):
            logger.warning("No label file extension defined.")
            return None
        label_path = pathlib.Path(self.file_path).with_suffix(self.file_wrapper_config["label_file_extension"])
        label = self.filesystem_wrapper.get(str(label_path))
        if label is not None:
            label = label.decode("utf-8")
            return int(label)
        return None

    def get_samples_from_indices(self, indices: list) -> bytes:
        """Get the samples from the file.

        Args:
            indices (list): Indices

        Raises:
            IndexError: If the indices are not valid

        Returns:
            bytes: Samples
        """
        if len(indices) != 1 or indices[0] != 0:
            raise IndexError("SingleSampleFileWrapper contains only one sample.")
        return self.get_sample(0)

File: modyn/storage/src/internal/file_wrapper/single_sample_file_wrapper.cpp
#include "internal/file_wrapper/single_sample_file_wrapper.hpp"

#include <spdlog/spdlog.h>

#include <filesystem>
#include <iostream>

#include "modyn/utils/utils.hpp"

using namespace modyn::storage;

uint64_t SingleSampleFileWrapper::get_number_of_samples() {
  ASSERT(file_wrapper_config_["file_extension"], "File wrapper configuration does not contain a file extension");
  const auto file_extension = file_wrapper_config_["file_extension"].as<std::string>();
  if (file_path_.find(file_extension) == std::string::npos) {
    return 0;
  }
  return 1;
}

int64_t SingleSampleFileWrapper::get_label(uint64_t /* index */) {
  ASSERT(file_wrapper_config_["file_extension"], "File wrapper configuration does not contain a label file extension");
  const auto label_file_extension = file_wrapper_config_["label_file_extension"].as<std::string>();
  auto label_path = std::filesystem::path(file_path_).replace_extension(label_file_extension);
  ASSERT(filesystem_wrapper_->exists(label_path), fmt::format("Label file does not exist: {}", label_path.string()));
  std::vector<unsigned char> label = filesystem_wrapper_->get(label_path);
  if (!label.empty()) {
    auto label_str = std::string(reinterpret_cast<char*>(label.data()), label.size());
    return std::stoi(label_str);
  }
  FAIL(fmt::format("Label file is empty: {}", label_path.string()));
  return -1;
}

std::vector<int64_t> SingleSampleFileWrapper::get_all_labels() { return std::vector<int64_t>{get_label(0)}; }

std::vector<unsigned char> SingleSampleFileWrapper::get_sample(uint64_t index) {
  ASSERT(index == 0,
         fmt::format("Single sample file wrappers can only access the first sample. file_path = {}, index = {}",
                     file_path_, index));
  return filesystem_wrapper_->get(file_path_);
}

std::vector<std::vector<unsigned char>> SingleSampleFileWrapper::get_samples(uint64_t start, uint64_t end) {
  ASSERT(
      start == 0 && end == 1,
      fmt::format("Single sample file wrappers can only access the first sample. file_path = {}, start = {}, end = {}",
                  file_path_, start, end));
  return std::vector<std::vector<unsigned char>>{get_sample(0)};
}

std::vector<std::vector<unsigned char>> SingleSampleFileWrapper::get_samples_from_indices(
    const std::vector<uint64_t>& indices) {
  ASSERT(indices.size() == 1 && indices[0] == 0,
         fmt::format("Single sample file wrappers can only access the first sample. file_path = {}, indices.size() = "
                     "{}, indices = [{}]",
                     file_path_, indices.size(), fmt::join(indices, ", ")));
  return std::vector<std::vector<unsigned char>>{get_sample(0)};
}

/*
 * get_targets for single sample file wrappers.
 * Reads from a separate target file whose extension is specified in the configuration as "target_file_extension".
 */
std::vector<std::vector<unsigned char>> SingleSampleFileWrapper::get_targets(uint64_t start, uint64_t end) {
  ASSERT(
      start == 0 && end == 1,
      fmt::format(
          "Single sample file wrappers can only access the first sample target. file_path = {}, start = {}, end = {}",
          file_path_, start, end));
  return std::vector<std::vector<unsigned char>>{get_target(0)};
}

/*
 * get_targets_from_indices for single sample file wrappers.
 */
std::vector<std::vector<unsigned char>> SingleSampleFileWrapper::get_targets_from_indices(
    const std::vector<uint64_t>& indices) {
  ASSERT(indices.size() == 1 && indices[0] == 0,
         fmt::format("Single sample file wrappers can only access the first sample target. file_path = {}, "
                     "indices.size() = {}, indices = [{}]",
                     file_path_, indices.size(), fmt::join(indices, ", ")));
  return std::vector<std::vector<unsigned char>>{get_target(0)};
}

void SingleSampleFileWrapper::validate_file_extension() {
  ASSERT(file_wrapper_config_["file_extension"], "File wrapper configuration does not contain a file extension");
  const auto file_extension = file_wrapper_config_["file_extension"].as<std::string>();
  if (file_path_.find(file_extension) == std::string::npos) {
    FAIL(fmt::format("File extension {} does not match file path {}", file_extension, file_path_));
  }
}

void SingleSampleFileWrapper::delete_samples(const std::vector<uint64_t>& /* indices */) {
  // The file will be deleted at a higher level
}

FileWrapperType SingleSampleFileWrapper::get_type() { return FileWrapperType::SINGLE_SAMPLE; }

/*
 * get_target for single sample file wrappers.
 * Reads from a separate target file whose extension is specified in the configuration as "target_file_extension".
 */
std::vector<unsigned char> SingleSampleFileWrapper::get_target(uint64_t index) {
  ASSERT(index == 0,
         fmt::format("Single sample file wrappers can only access the first sample. file_path = {}, index = {}",
                     file_path_, index));
  ASSERT(file_wrapper_config_["target_file_extension"],
         "File wrapper configuration does not contain a target file extension");
  const auto target_file_extension = file_wrapper_config_["target_file_extension"].as<std::string>();
  auto target_path = std::filesystem::path(file_path_).replace_extension(target_file_extension);
  ASSERT(filesystem_wrapper_->exists(target_path), fmt::format("Target file does not exist: {}", target_path.string()));
  return filesystem_wrapper_->get(target_path);
}
